Custom objdump support, formatting adjustments to assembly in debug output, and minor fixes

karas84 opened the pull request:
This PR introduces the following changes:

## FIXES
- Escaped `%` in the help string to prevent argparse help message generation from breaking.
- Set speed value to 100 when creating permuter instance in evaluator.py
  (parameter was not added after the signature was changed when the speed parameter was introduced)

## FEATURES
- Added support for specifying a custom `objdump` executable path and custom command-line options.  
  (These settings can be configured in `permuter_settings.toml` and `settings.toml`. When importing settings, they are automatically transferred to `settings.toml`)
- Added a command-line option to disable ignoring branch targets.  
  (The `ign_branch_targets` variable has been replaced with a `DEFAULT_IGN_BRANCH_TARGETS` constant, which serves as both a default function parameter and the basis for generating the command-line argument:  
  - `--no-ignore-branch-targets` if `DEFAULT_IGN_BRANCH_TARGETS` is `TRUE`  
  - `--ignore-branch-targets` if `FALSE`)
- Improved assembly formatting in debug mode by ensuring proper alignment.  
  (Instead of relying on tabs for formatting—potentially causing misaligned columns—the code now calculates the maximum mnemonic length and aligns all mnemonics accordingly. The 40-character column limit for target assembly remains unchanged.)
- Removed color highlighting for lines where the new field matches a symbol but the old field contained a relocation.  
  (For example, `ld a0,%lo(D_0034CFF0)(v0)` <-> `ld a0,%lo(.rodata)(v0)`. These cases are not internally counted as differences, and decomp.me does not color these diffs either, making this behavior more consistent.)

## CHORES
- Improved type hinting to comply with stricter type checkers (e.g., Pyright in strict mode).  
  (For type annotations of basic types such as tuples, lists, and dictionaries, the corresponding types from the `typing` module are now used instead of built-in types to maintain backward compatibility with older Python versions.)
- Added common virtualenv paths to `.gitignore`
  (`/.venv/` and `/venv/`)

simonlindholm commented on src/main.py:
```diff
@@ -600,7 +613,7 @@ def __call__(
         option_string: Optional[str] = None,
     ) -> None:
-        weights = get_default_randomization_weights("base")
+        _weights = get_default_randomization_weights("base")
```
oops!
```suggestion
        # TODO include weights from get_default_randomization_weights("base")
```

karas84 replied:
I wasn't sure I could remove / comment it hehe.

simonlindholm reviewed (commented):
It would be nice if specifying a custom objdump wasn't needed. Would it be possible to specify -m mips:5900 unconditionally, or does it break things on older mips? Or could we auto-detect what flags to use based on the ELF?

karas84 commented on import.py:
```diff
@@ -292,8 +303,8 @@ def find_build_command_line(
-    output = []
-    close_match = ""
+    output: List[List[str]] = []
+    close_match: str = ""
```
The only really "necessary" is the list annotation. When you declare a list without annotating it, Pyright considers it as `list[Unknown]` thus, in strict mode, all uses of that list are marked in red. for example, in the `output.append(cmdline)` few lines later, I get this from Pyright:
```python
Type of "append" is partially unknown
  Type of "append" is "(object: Unknown, /) -> None"Pylance[reportUnknownMemberType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportUnknownMemberType)
(method) def append(
    object: Unknown,
    /
) -> None
```

I guess that enforcing a specific type ensures that you are not, by mistake, appending something (or doing operations) with the wrong type.

The typing on the string is something I usually do for all variables. If you don't specify a type, the type checker assigns one based on how you use the variable, and if you use the same variable with different types the final type will be an aggregation of those types (e.g. `str | int | None`). By specifying a type you force the type checker to check for wrong type usages (suppose you accidentally assign an int to that var later, you then get an error).

This is a consequence of using strict type checking. If you are fine with the base/standard one, we don't need this.

simonlindholm replied:
I'm happy with strict typing, and explicitly typing lists feels like it helps readability slightly too. (I think mypy infers types for them, or else I would have noticed something.) I was just curious about the explicitly typed `str` variable. I guess the difference there is that mypy would never infer `str | int | None`, but rather just go with the originally assigned type. I wonder if pyright can be configured the same way. `: str` here is kinda cluttery imo, but I don't actually care too much.

karas84 commented on src/main.py:
```diff
@@ -763,11 +776,25 @@ def main() -> None:
+            "--ignore-branch-targets",
+            dest="ign_branch_targets",
+            action="store_true",
+            help="Ignore branch targets in objdump output",
```
Sure, that makes sense.

karas84 commented on src/main.py:
```diff
@@ -763,11 +776,25 @@ def main() -> None:
         metavar="[1-100]",
         default=100,
     )
+    if DEFAULT_IGN_BRANCH_TARGETS:
```
I totally agree. The rationale for that is that now the behaviour is harcoded and all targets are removed, and the comment you left explains very well why is that, so I wasn't sure changing that would be acceptable. The default value thing is a compromise I did for the PR (in my local code I just added an extra param and that value is propagated in the code). If you are OK with the change, I'd suggest we do it like in my local repo: remove the default and add a standard argument that has as default the current behaviour, making it something similar to the `--stack-diffs` flag. So by default the targets would still be removed, but if the user wants, they can enable them.

simonlindholm replied:
I think that's good, yes. The comment could move into the cmdline help text maybe.

karas84 commented on src/net/core.py:
```diff
@@ -61,7 +71,7 @@ class PermuterData:
 
 def permuter_data_from_json(
-    obj: dict, source: str, target_o_bin: bytes
+    obj: Dict[Any, Any], source: str, target_o_bin: bytes
```
Yeah I got the Idea :) I'll fix them all as you suggested.

karas84 commented on src/scorer.py:
```diff
@@ -23,20 +23,33 @@ def __init__(
         debug_mode: bool,
+        objdump_path: Optional[str] = None,
+        objdump_args: Optional[List[str]] = None,
+        ign_branch_targets: bool = DEFAULT_IGN_BRANCH_TARGETS,
```
Didn't want to break any external library using that code. But if you are OK with that I'll remove the defaults.

karas84 replied:
Removed the defaults. To do that I also had to add those to PermuterData and make them mandatory in the evaluator (unless you want the evaluator to always use the default objdump with default args, but that would be inconsistent with the rest of the program). I also updated `permuter_data_from_json` and `permuter_data_to_json` accordingly. Don't know if this was your intention. If you don't like it I can revert the changes.

karas84 commented on src/scorer.py:
```diff
@@ -171,20 +195,52 @@ def remap(seq: List[str]) -> str:
         if self.debug_mode:
+            # find the max mnemonic length for consistent padding
+            mnem_max_len = max(
+                max(len(line.mnemonic) for line in self.target_seq),
+                max(len(line.mnemonic) for line in cand_seq),
```
Didn't think about that case hehe. But is it really possible? Would that mean that either the target or the candidate ASM is empty? Anyway, better safe than sorry. I'll fix that.

simonlindholm replied:
Yeah I don't think it's actually possible, it's just general cleanliness

karas84 commented on src/scorer.py:
```diff
@@ -157,12 +179,14 @@ def remap(seq: List[str]) -> str:
             result_diff = self.difflib_differ.get_opcodes()
 
-        for (tag, i1, i2, j1, j2) in result_diff:
+        ignore_diff: List[Line] = []
```
I actually implemented it with a `Set[int]` first, then changed it to how it is now because it was cleaner. The way I did that was to use a counter variable and increment it in each of the 3 tag loops (as I don't think there is a direct way to know the line number), then reset the counter and do the same in the debug loops to check against the indices in the set. But yeah, I should have used a `tuple` `(old_line, new_line)`, instead of just the new line since that would have side effects like you mentioned. I can implement it both ways, what do you prefer?

karas84 commented on src/scorer.py:
```diff
@@ -171,20 +195,52 @@ def remap(seq: List[str]) -> str:
-                        color = "\u001b[0m" if old == new else "\u001b[94m"
-                        print(color, old[:40].ljust(40), "\t", new)
+                        new = new_line.row
+                        same = old == new or new_line in ignore_diff
```
Yeah I guess not :D

karas84 commented:
> It would be nice if specifying a custom objdump wasn't needed. Would it be possible to specify -m mips:5900 unconditionally, or does it break things on older mips? Or could we auto-detect what flags to use based on the ELF?

I agree, not needing to do anything would indeed be desirable. Can it really be avoided though? I don't think I have the right amount of knowledge to answer this, maybe we should discuss it in the server with the decompal guys.

I'm pretty sure using `-m mips:5900` with PSX games would break something, and also I had to use the custom decompal PS2 objdump for RE: Code Veronica as the default one with the same exact parameters did not work, so I fear it may be not just a question of parameters.

karas84 commented:
While making the changes you suggested, I realized that my approach to enabling branch differences was too naive. I was displaying the differences, but I wasn’t counting them appropriately. To address this, I added a penalty score for branch differences, similar to the stack differences, but based only on their count.

I also added an extra check in `diff_sameline` to identify branch differences and increment the branch difference penalty counter. This counter is now factored into the final diff score calculation. The debug output also displays the branch penalty.

simonlindholm commented on src/main.py:
```diff
@@ -545,7 +558,7 @@ def get_task(perm_index: int) -> Optional[Tuple[int, int]]:
                     if options.stop_on_zero:
                         break
-            elif isinstance(feedback, NeedMoreWork):
+            elif isinstance(feedback, NeedMoreWork):  # pyright: ignore[reportUnnecessaryIsInstance]
```
I'd prefer per-IDE settings to lint comments in the code.

simonlindholm commented on src/net/core.py:
```diff
@@ -77,6 +80,9 @@ def permuter_data_from_json(
         target_o_bin=target_o_bin,
+        ign_branch_targets=json_prop(obj, "ign_branch_targets", bool),
+        objdump_path=json_prop(obj, "objdump_path", str),
+        objdump_args=json_prop(obj, "objdump_args", str),
```
Here it makes sense to default to True / "" / "" for now to preserve compatibility with old clients

simonlindholm commented on src/net/core.py:
```diff
@@ -57,6 +57,9 @@ class PermuterData:
     source: str
     target_o_bin: bytes
+    ign_branch_targets: bool
+    objdump_path: Optional[str]
+    objdump_args: Optional[str]
```
these are optional while the constructor always passes str

karas84 commented:
> It would be nice if specifying a custom objdump wasn't needed. Would it be possible to specify -m mips:5900 unconditionally, or does it break things on older mips? Or could we auto-detect what flags to use based on the ELF?

Still don't know if the first part is possible, but I've created a gist to try to find the proper architecture from the ELF:

https://gist.github.com/karas84/5016f9c873e2a219965a4d561e2bef9c

For now it only checks for the machine type and some flags to infer the architecture.

karas84 commented:
Last commit merges `objdump_path` and `objdump_args` into `objdump_command`

simonlindholm commented:
Sorry for the long delay!

simonlindholm commented:
What's the reason for reverting
```diff
-    compiler_mappings = cast(Dict[str, object], compiler_mappings)
+    compiler_mappings = cast(Dict[str, str], compiler_mappings)
```
btw? I tried downloading pyright but I'm getting spamming by errors, so I might be doing something wrong... It seems like we at the very least need a pyproject.toml with
```toml
[tool.pyright]
stubPath = "stubs/"
```
(which would be nice to commit), but I guess your setup is more involved than that.

karas84 commented:
> Sorry for the long delay!

no problem at all!


`compiler_mappings` is read from `permuter_settings.toml` and is a mapping between a local compiler path and the name of a compiler on decomp.me, something like:

```
[decompme.compilers]
"/local/path/to/compilers/ps2/mwcps2-2.3.3-000906/mwccps2" = "mwcps2-2.3.3-000906"
...
```
so it's a str-str mapping. Also in the code, some lines below that, there's:
```
    for path, compiler_name in compiler_mappings.items():
        assert isinstance(compiler_name, str)
```
The intent was to give it the correct type from the start, unless you want to remain generic (object values) and check the type of each value in the loop? The decision is up to you of course.

simonlindholm commented:
I see. I'd say that if you type it as str->str from the start the type is a lie, because it's not been verified yet that that is actually correct (users can put whatever in their configs). That's the role of the isinstance check. I pushed a follow-up commit that changes it back to str->object.

karas84 commented:
About pyright, I have it integrated in vscode so I didn't have to setup it, it "just works". The only change I made was to enable strict typing in the settings: Settings > Extensions > Pylance > Python › Analysis: Type Checking Mode > Strict, or by hand in the `.vscode/settings.json`: `"python.analysis.typeCheckingMode": "strict"`.

> btw? I tried downloading pyright but I'm getting spamming by errors, so I might be doing something wrong... It seems like we at the very least need a pyproject.toml with ...

Great, I had to point vscode to the stubs folder in the settings for it to read them. Didn't want to change that part myself.

karas84 commented:
> I see. I'd say that if you type it as str->str from the start the type is a lie, because it's not been verified yet that that is actually correct (users can put whatever in their configs). That's the role of the isinstance check. I pushed a follow-up commit that changes it back to str->object.

Makes sense.

I'm seeing this in the main branch:
```
    decompme_settings = json_dict(settings, "decompme")
    compiler_mappings = json_dict(decompme_settings, "compilers")

    compiler_path = compiler[0]
```
instead of
```
    decompme_settings = settings.get("decompme", {})
    assert isinstance(decompme_settings, dict)
    decompme_settings = cast(Dict[str, object], decompme_settings)
    compiler_mappings = decompme_settings.get("compilers", {})
    assert isinstance(compiler_mappings, dict)
    compiler_mappings = cast(Dict[str, object], compiler_mappings)

    compiler_path = compiler[0]
```
is that intended? (It's not an issue, just curious, maybe it was too much hehe)

karas84 commented:
Oh I see, you made a new `json_dict` function to clean things up. nice!
